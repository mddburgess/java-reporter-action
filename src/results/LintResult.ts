import { countBy, Dictionary, groupBy, toPairs } from "lodash";
import CheckResult from "../check/result";
import { plural } from "../common/utils";
import { CheckAnnotation, CheckConclusion } from "../github/types";
import LintAnnotation, { toCheckAnnotation } from "./LintAnnotation";

export default class LintResult extends CheckResult {
  private readonly countByLevel: Dictionary<number>;

  public constructor(private readonly lintAnnotations: LintAnnotation[]) {
    super();
    this.countByLevel = countBy(this.lintAnnotations, (annotation) => annotation.level);
  }

  public shouldCompleteCheck(): boolean {
    return true;
  }

  public get conclusion(): CheckConclusion {
    if (this.countByLevel["failure"] > 0) {
      return "failure";
    }
    if (this.countByLevel["warning"] > 0) {
      return "neutral";
    }
    return "success";
  }

  public get title(): string {
    if (this.countByLevel["failure"] > 0) {
      return plural(this.countByLevel["failure"], "failure");
    }
    if (this.countByLevel["warning"] > 0) {
      return plural(this.countByLevel["warning"], "warning");
    }
    if (this.countByLevel["notice"] > 0) {
      return plural(this.countByLevel["notice"], "notice");
    }
    return "Passed";
  }

  public get summary(): string {
    if (this.lintAnnotations.length === 0) {
      return "Passed";
    }

    const categories = toPairs(groupBy(this.lintAnnotations, (annotation) => annotation.category))
      .map(([category, annotations]) => ({
        category: category,
        failures: annotations.filter((a) => a.level === "failure").length,
        warnings: annotations.filter((a) => a.level === "warning").length,
        notices: annotations.filter((a) => a.level === "notice").length,
      }))
      .sort((a, b) => a.category.localeCompare(b.category))
      .map((o) => `| ${o.category} | ${o.failures} | ${o.warnings} | ${o.notices} |`);

    return [
      "| Category | Failures | Warnings | Notices |",
      "| :-- | --: | --: | --: |",
      ...categories,
    ].join("\n");
  }

  public get text(): string | undefined {
    if (this.lintAnnotations.length === 0) {
      return undefined;
    }

    const paths = toPairs(groupBy(this.lintAnnotations, (annotation) => annotation.path))
      .map(([path, annotations]) => ({
        path: path,
        failures: annotations.filter((a) => a.level === "failure").length,
        warnings: annotations.filter((a) => a.level === "warning").length,
        notices: annotations.filter((a) => a.level === "notice").length,
      }))
      .sort((a, b) => a.path.localeCompare(b.path))
      .map((o) => `| \`${o.path}\` | ${o.failures} | ${o.warnings} | ${o.notices} |`);

    return ["| Path | Failures | Warnings | Notices |", "| :-- | --: | --: | --: |", ...paths].join(
      "\n"
    );
  }

  public get annotations(): CheckAnnotation[] {
    return this.lintAnnotations.map(toCheckAnnotation);
  }
}
