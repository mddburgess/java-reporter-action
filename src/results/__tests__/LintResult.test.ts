import LintResult from "../LintResult";

describe("LintResult", () => {
  it("reports success for an empty annotation list", () => {
    const result = new LintResult([]);

    expect(result.shouldCompleteCheck()).toBe(true);
    expect(result.conclusion).toBe("success");
    expect(result.title).toBe("Passed");
    expect(result.summary).toBe("Passed");
    expect(result.text).toBeUndefined();
    expect(result.annotations).toStrictEqual([]);
  });

  it("reports success for an annotation list containing notices", () => {
    const result = new LintResult([
      {
        path: "path/to/Notice.java",
        line: 1,
        level: "notice",
        category: "category 1",
        type: "type",
        message: "message",
        details: "details",
      },
    ]);

    expect(result.shouldCompleteCheck()).toBe(true);
    expect(result.conclusion).toBe("success");
    expect(result.title).toBe("1 notice");
    expect(result.summary).toMatchSnapshot();
    expect(result.text).toMatchSnapshot();
    expect(result.annotations).toMatchSnapshot();
  });

  it("reports neutral for an annotation list containing warnings", () => {
    const result = new LintResult([
      {
        path: "path/to/Notice.java",
        line: 1,
        level: "notice",
        category: "category 1",
        type: "type",
        message: "message",
        details: "details",
      },
      {
        path: "path/to/Warning.java",
        line: 1,
        level: "warning",
        category: "category 2",
        type: "type",
        message: "message",
        details: "details",
      },
    ]);

    expect(result.shouldCompleteCheck()).toBe(true);
    expect(result.conclusion).toBe("neutral");
    expect(result.title).toBe("1 warning");
    expect(result.summary).toMatchSnapshot();
    expect(result.text).toMatchSnapshot();
    expect(result.annotations).toMatchSnapshot();
  });

  it("reports failure for an annotation list containing failures", () => {
    const result = new LintResult([
      {
        path: "path/to/Notice.java",
        line: 1,
        level: "notice",
        category: "category 1",
        type: "type",
        message: "message",
        details: "details",
      },
      {
        path: "path/to/Warning.java",
        line: 1,
        level: "warning",
        category: "category 2",
        type: "type",
        message: "message",
        details: "details",
      },
      {
        path: "path/to/Failure.java",
        line: 1,
        level: "failure",
        category: "category 3",
        type: "type",
        message: "message",
        details: "details",
      },
    ]);

    expect(result.shouldCompleteCheck()).toBe(true);
    expect(result.conclusion).toBe("failure");
    expect(result.title).toBe("1 failure");
    expect(result.summary).toMatchSnapshot();
    expect(result.text).toMatchSnapshot();
    expect(result.annotations).toMatchSnapshot();
  });
});
