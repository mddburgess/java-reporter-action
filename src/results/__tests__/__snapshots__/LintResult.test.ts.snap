// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LintResult reports failure for an annotation list containing failures 1`] = `
"| Category | Failures | Warnings | Notices |
| :-- | --: | --: | --: |
| category 1 | 0 | 0 | 1 |
| category 2 | 0 | 1 | 0 |
| category 3 | 1 | 0 | 0 |"
`;

exports[`LintResult reports failure for an annotation list containing failures 2`] = `
"| Path | Failures | Warnings | Notices |
| :-- | --: | --: | --: |
| \`path/to/Failure.java\` | 1 | 0 | 0 |
| \`path/to/Notice.java\` | 0 | 0 | 1 |
| \`path/to/Warning.java\` | 0 | 1 | 0 |"
`;

exports[`LintResult reports failure for an annotation list containing failures 3`] = `
Array [
  Object {
    "annotation_level": "notice",
    "end_line": 1,
    "message": "message",
    "path": "path/to/Notice.java",
    "raw_details": "details",
    "start_line": 1,
    "title": "category 1: type",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 1,
    "message": "message",
    "path": "path/to/Warning.java",
    "raw_details": "details",
    "start_line": 1,
    "title": "category 2: type",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 1,
    "message": "message",
    "path": "path/to/Failure.java",
    "raw_details": "details",
    "start_line": 1,
    "title": "category 3: type",
  },
]
`;

exports[`LintResult reports neutral for an annotation list containing warnings 1`] = `
"| Category | Failures | Warnings | Notices |
| :-- | --: | --: | --: |
| category 1 | 0 | 0 | 1 |
| category 2 | 0 | 1 | 0 |"
`;

exports[`LintResult reports neutral for an annotation list containing warnings 2`] = `
"| Path | Failures | Warnings | Notices |
| :-- | --: | --: | --: |
| \`path/to/Notice.java\` | 0 | 0 | 1 |
| \`path/to/Warning.java\` | 0 | 1 | 0 |"
`;

exports[`LintResult reports neutral for an annotation list containing warnings 3`] = `
Array [
  Object {
    "annotation_level": "notice",
    "end_line": 1,
    "message": "message",
    "path": "path/to/Notice.java",
    "raw_details": "details",
    "start_line": 1,
    "title": "category 1: type",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 1,
    "message": "message",
    "path": "path/to/Warning.java",
    "raw_details": "details",
    "start_line": 1,
    "title": "category 2: type",
  },
]
`;

exports[`LintResult reports success for an annotation list containing notices 1`] = `
"| Category | Failures | Warnings | Notices |
| :-- | --: | --: | --: |
| category 1 | 0 | 0 | 1 |"
`;

exports[`LintResult reports success for an annotation list containing notices 2`] = `
"| Path | Failures | Warnings | Notices |
| :-- | --: | --: | --: |
| \`path/to/Notice.java\` | 0 | 0 | 1 |"
`;

exports[`LintResult reports success for an annotation list containing notices 3`] = `
Array [
  Object {
    "annotation_level": "notice",
    "end_line": 1,
    "message": "message",
    "path": "path/to/Notice.java",
    "raw_details": "details",
    "start_line": 1,
    "title": "category 1: type",
  },
]
`;
